### PR TITLE
CLOUD-215 Add IO functions to read and write text of type string

### PR DESCRIPTION
**Pull request by ibadami**
- Converts all the three functions Dump, readHeader and ReadModel to DumpText, readHeaderText and ReadModelFromText

### CODE REVIEW

**Comment by stelioskat** [comment]
I think that it is more fitting to use a bit different naming scheme for the new functions. Instead of using the word "Text" use "String" E.g. `NewModelFromText` might be better to be called `NewModelFromString`.

Furthermore, it would be nice to also add documentation to all fun exported elements, at least for files `model.go` and `modelio.go` to make a start :)

File: modelio.go [hunk 1]
```diff
@@ -340,3 +340,301 @@ func (model *Model) ReadModel(file string) error {
...
+	if model.param.KernelType == POLY || model.param.KernelType == SIGMOID {
+		output = append(output, fmt.Sprintf("coef0 %.6g\n", model.param.Coef0))
+	}
+
+	var nrClass int = model.nrClass
+	output = append(output, fmt.Sprintf("nr_class %d\n", nrClass))
+
+	var l int = model.l
+	output = append(output, fmt.Sprintf("total_sv %d\n", l))
+
```
Comment (sertopexamgio) [inline modelio.go]: Don't we need `if len(model.rho) > 0 {` here ?

File: modelio.go [hunk 2]
```diff
@@ -340,3 +340,301 @@ func (model *Model) ReadModel(file string) error {
...
+				index := model.svSpace[i_idx].index
+				value := model.svSpace[i_idx].value
+				output = append(output, fmt.Sprintf("%d:%.8g ", index, value))
+				i_idx++
+			}
+			output = append(output, "\n")
+		}
+	}
+
+	strings.Join(output, "")
```
Comment (sertopexamgio) [inline modelio.go]: Do we need this line? Because we have it again on the next line.

File: modelio.go [hunk 3]
```diff
@@ -340,3 +340,301 @@ func (model *Model) ReadModel(file string) error {
...
+				if model.probA[i], err = strconv.ParseFloat(tokens[i+1], 64); err != nil {
+					return lineNumber, err
+				}
+			}
+
+		case "probB":
+
+			total_class_comparisons := model.nrClass * (model.nrClass - 1) / 2
+			if total_class_comparisons != len(tokens)-1 {
+				return lineNumber, fmt.Errorf("Number of probB %d does not mactch the required number %d\n", len(tokens)-1, total_class_comparisons)
```
Comment (sertopexamgio) [inline modelio.go]: typo: `match`

File: modelio.go [hunk 4]
```diff
@@ -340,3 +340,301 @@ func (model *Model) ReadModel(file string) error {
...
+
+func (model *Model) readHeaderText(text []string) (int, error) {
+
+	var lineNumber int
+
+	for l, line := range text {
+		var i = 0
+		var err error
+
+		tokens := strings.Fields(line)
```
Comment (sertopexamgio) [inline modelio.go]: Thus we use `line` only here, a small optimization could be to use here text[l] and erase the variable line from the `for statement`.

**Comment by ibadami** [comment]
Great reviews guys, thanks!

**Comment by ibadami** [comment]
No, SV should be there. If there are no Support vectors after the SV entry, the readingFunction will throw error. But reading function must read until SV.

**Pull request merged**